### PR TITLE
Support unqualified and partially qualified domain name in DNS query in Windows kube-proxy

### PR DESCRIPTION
**Pull request by JiangtianLi**
**What this PR does / why we need it**:
In Windows container networking, --dns-search is not currently supported on Windows Docker. Besides, even with --dns-suffix, inside Windows container DNS suffix is not appended to DNS query names. That makes unqualified domain name or partially qualified domain name in DNS query not able to resolve.

This PR provides a solution to resolve unqualified domain name or partially qualified domain name in DNS query for Windows container in Windows kube-proxy. It uses well-known Kubernetes DNS suffix as well host DNS suffix search list to append to the name in DNS query. DNS packet in kube-proxy UDP stream is modified as appropriate.

This PR affects the Windows kube-proxy only.

**Special notes for your reviewer**:
This PR is based on top of Anthony Howe's commit 48647fb, 0e37f0a and 7e2c71f which is already included in the PR 41487. Please only review commit b9dfb69.

**Release note**:

```release-note
Add DNS suffix search list support in Windows kube-proxy.
```

### CODE REVIEW

**Comment by JiangtianLi** [comment]
I signed it! Thanks.

File: pkg/proxy/winuserspace/proxier.go [hunk 1]
```diff
@@ -28,7 +27,7 @@ import (
 	"github.com/golang/glog"
 
 	"k8s.io/apimachinery/pkg/types"
-	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	//utilerrors "k8s.io/apimachinery/pkg/util/errors"
```
Comment (brendandburns) [inline pkg/proxy/winuserspace/proxier.go]: Delete this if its no longer used.

File: pkg/proxy/winuserspace/proxier.go [hunk 2]
```diff
@@ -37,25 +36,26 @@ import (
...
-	loadBalancerStatus  api.LoadBalancerStatus
+	isAliveAtomic int32 // Only access this with atomic ops
+	portal        portal
+	protocol      api.Protocol
+	proxyPort     int
+	socket        proxySocket
+	timeout       time.Duration
+	activeClients *clientCache
+	dnsClients    *dnsClientCache
+	//nodePort      int
```
Comment (brendandburns) [inline pkg/proxy/winuserspace/proxier.go]: delete these if the are no longer used.

File: pkg/proxy/winuserspace/proxier.go [hunk 3]
```diff
@@ -85,16 +85,16 @@ func logTimeout(err error) bool {
...
 	syncPeriod     time.Duration
 	udpIdleTimeout time.Duration
 	portMapMutex   sync.Mutex
 	portMap        map[portMapKey]*portMapValue
 	numProxyLoops  int32 // use atomic ops to access this; mostly for testing
-	listenIP       net.IP
-	netsh          netsh.Interface
-	hostIP         net.IP
-	proxyPorts     PortAllocator
+	//listenIP       net.IP
```
Comment (brendandburns) [inline pkg/proxy/winuserspace/proxier.go]: here too.

File: pkg/proxy/winuserspace/proxier.go [hunk 4]
```diff
@@ -85,16 +85,16 @@ func logTimeout(err error) bool {
...
+	netsh  netsh.Interface
+	hostIP net.IP
+	//proxyPorts     PortAllocator
```
Comment (brendandburns) [inline pkg/proxy/winuserspace/proxier.go]: and here

File: pkg/proxy/winuserspace/proxier.go [hunk 5]
```diff
@@ -196,60 +188,64 @@ func (proxier *Proxier) SyncLoop() {
 
 // Ensure that portals exist for all services.
 func (proxier *Proxier) ensurePortals() {
-	proxier.mu.Lock()
+	/*proxier.mu.Lock()
```
Comment (brendandburns) [inline pkg/proxy/winuserspace/proxier.go]: delete this block.

File: pkg/proxy/winuserspace/proxier.go [hunk 6]
```diff
@@ -196,60 +188,64 @@ func (proxier *Proxier) SyncLoop() {
...
 func (proxier *Proxier) cleanupStaleStickySessions() {
 	proxier.mu.Lock()
 	defer proxier.mu.Unlock()
-	for name := range proxier.serviceMap {
+	spnMap := make(map[proxy.ServicePortName]bool)
+	for k := range proxier.serviceMap {
+		spn := proxy.ServicePortName{NamespacedName: types.NamespacedName{Namespace: k.Namespace, Name: k.Name}, Port: k.Port}
+		spnMap[spn] = true
+	}
+	for name := range spnMap {
```
Comment (brendandburns) [inline pkg/proxy/winuserspace/proxier.go]: This indirection seems weird, why not just `proxier.loadBalancer.CleanupStaleStickySessions(name)` where you set `spnMap[spn]=true`

File: pkg/proxy/winuserspace/proxier.go [hunk 7]
```diff
@@ -304,84 +301,144 @@ func (proxier *Proxier) OnServiceUpdate(services []api.Service) {
...
-			info, exists := proxier.getServiceInfo(serviceName)
-			// TODO: check health of the socket?  What if ProxyLoop exited?
-			if exists && sameConfig(info, service, servicePort) {
-				// Nothing changed.
-				continue
+			// create a slice of source IPs
+			var listenIPs []string
+			listenIPs = append(listenIPs, service.Spec.ClusterIP)
+
+			/*for _, ip := range service.Spec.ExternalIPs {
```
Comment (brendandburns) [inline pkg/proxy/winuserspace/proxier.go]: delete dead code.

File: pkg/proxy/winuserspace/proxier.go [hunk 8]
```diff
@@ -304,84 +301,144 @@ func (proxier *Proxier) OnServiceUpdate(services []api.Service) {
...
-				err = proxier.stopProxy(serviceName, info)
-				if err != nil {
-					glog.Errorf("Failed to stop service %q: %v", serviceName, err)
+				if exists {
+					glog.V(4).Infof("Something changed for service %q: stopping it", serviceName)
+
+					// turn off the proxy
+					err := proxier.stopProxy(serviceName, info)
+					if err != nil {
+						glog.Errorf("Failed to stop service %q: %v", serviceName, err)
```
Comment (brendandburns) [inline pkg/proxy/winuserspace/proxier.go]: bail out here?

File: pkg/proxy/winuserspace/proxier.go [hunk 9]
```diff
@@ -304,84 +301,144 @@ func (proxier *Proxier) OnServiceUpdate(services []api.Service) {
...
+					}
+
+					// close the PortalProxy if it is not a node port
+					if serviceIP != nil {
+						args := proxier.netshIpv4AddressDeleteArgs(serviceIP)
+						if err := proxier.netsh.DeleteIPAddress(args); err != nil {
+							glog.Errorf("Failed to delete IP address for service %q, error %s", serviceName, err.Error())
+						}
+					} else {
+						// TODO(ajh) release the node port
```
Comment (brendandburns) [inline pkg/proxy/winuserspace/proxier.go]: Address this todo?

File: pkg/proxy/winuserspace/proxier.go [hunk 10]
```diff
@@ -304,84 +301,144 @@ func (proxier *Proxier) OnServiceUpdate(services []api.Service) {
...
+					existed, err := proxier.netsh.EnsureIPAddress(args, serviceIP)
+					if err != nil {
+						glog.Errorf("Failed to add ip address for service %q, args:%v", serviceName, args)
+						continue
+					}
+					if !existed {
+						glog.V(3).Infof("Added ip address to fowarder interface for service %q on %s %s:%d", serviceName, servicePort.Protocol, serviceIP, int(servicePort.Port))
+					}
+				} else {
+					// TODO(ajh) handle the node port
```
Comment (brendandburns) [inline pkg/proxy/winuserspace/proxier.go]: as above.

File: pkg/proxy/winuserspace/proxier.go [hunk 11]
```diff
@@ -304,84 +301,144 @@ func (proxier *Proxier) OnServiceUpdate(services []api.Service) {
...
-				continue
+				// add the listener, proxy
+				sock, err := newProxySocket(servicePort.Protocol, serviceIP, int(servicePort.Port))
+				if err != nil {
+					glog.Errorf("failed to create a new proxy socket for service %q: %v", serviceName, err)
+					continue
+				}
+				si := &serviceInfo{
+					isAliveAtomic:       1,
+					portal:              portal{ip: listenIP, port: int(servicePort.Port), isExternal: false},
```
Comment (brendandburns) [inline pkg/proxy/winuserspace/proxier.go]: Multi-line initializer.

File: pkg/proxy/winuserspace/proxier.go [hunk 12]
```diff
@@ -304,84 +301,144 @@ func (proxier *Proxier) OnServiceUpdate(services []api.Service) {
...
+				}
+			} else {
+				// TODO(ajh) release the node port
+			}
+		}
+	}
+	// only delete spn if all listen ips show inactive
+	for k := range spnMap {
+		if !spnMap[k] {
+			proxier.loadBalancer.DeleteService(k)
```
Comment (brendandburns) [inline pkg/proxy/winuserspace/proxier.go]: As above seems like you could inline this in the previous loop.

File: pkg/proxy/winuserspace/proxier.go [hunk 13]
```diff
@@ -304,84 +301,144 @@ func (proxier *Proxier) OnServiceUpdate(services []api.Service) {
...
 		}
 	}
 }
 
 func sameConfig(info *serviceInfo, service *api.Service, port *api.ServicePort) bool {
-	if info.protocol != port.Protocol || info.portal.port != int(port.Port) || info.nodePort != int(port.NodePort) {
-		return false
-	}
-	if !info.portal.ip.Equal(net.ParseIP(service.Spec.ClusterIP)) {
```
Comment (brendandburns) [inline pkg/proxy/winuserspace/proxier.go]: Why did you delete all of this?

File: pkg/proxy/winuserspace/proxier.go [hunk 14]
```diff
@@ -402,6 +459,7 @@ func ipsEqual(lhs, rhs []string) bool {
 	return true
 }
 
+/*
```
Comment (brendandburns) [inline pkg/proxy/winuserspace/proxier.go]: Delete commented out code.

File: pkg/proxy/winuserspace/proxysocket.go [hunk 15]
```diff
@@ -197,8 +231,399 @@ func newClientCache() *clientCache {
 	return &clientCache{clients: map[string]net.Conn{}}
 }
 
-func (udp *udpProxySocket) ProxyLoop(service proxy.ServicePortName, myInfo *serviceInfo, proxier *Proxier) {
+// TODO: use Go net dnsmsg library to walk DNS message format
+// DNS packet header
+type dnsHeader struct {
```
Comment (brendandburns) [inline pkg/proxy/winuserspace/proxysocket.go]: This code could really use some unit tests.

**Review by brendandburns — CHANGES_REQUESTED** [review]

File: pkg/proxy/winuserspace/proxier.go (hunk 1 above)
Comment (JiangtianLi) [inline pkg/proxy/winuserspace/proxier.go]: Sorry for the confusion. My commit is based on Anthony's commit 302fa44 and 1410e06 so I had to include his commits here. In previous PR, the second commit 1410e06 was lost when I merge his and squash my commits. Now I did rebase and it looked fine. Please skip reviewing diff in commit 302fa44 and 1410e06.

File: pkg/proxy/winuserspace/proxysocket.go (hunk 15 above)
Comment (JiangtianLi) [inline pkg/proxy/winuserspace/proxysocket.go]: Thanks for the suggestion. I will add unit tests.

Comment (JiangtianLi) [inline pkg/proxy/winuserspace/proxysocket.go]: I added unit tests in a new commit.

**Comment by brendandburns** [comment]
@k8s-bot bazel test this

**Comment by brendandburns** [comment]
@JiangtianLi can you add a descriptive title for this PR (e.g. "support dns search in windows proxy")

Thanks
--brendan

**Comment by JiangtianLi** [comment]
@brendandburns Thanks for the reminder. I have modified the title.

**Comment by michmike** [comment]
cc: @pires as well for a review

**Review by brendandburns — APPROVED** [review]

**Comment by brendandburns** [comment]
@k8s-bot kops aws e2e test this

**Pull request merged**